lcs/agents/xcs/ClassifiersList.py

```diff
@@ -0,0 +1,149 @@
+        self.cfg = cfg
+        super().__init__(*args, oktypes=oktypes)
+
+    def insert_in_population(self, cl: Classifier):
+        for c in self:
+            if c.condition == cl.condition and c.action == cl.action:
+                c.numerosity += 1
+                return 
+        self.append(cl)
+
+    def generate_covering_classifier(self, situation, action, time_stamp):
+        # both Perception and string has __getitem__
+        # this way allows situation to be either str or Perception
+        generalized = []
+        for i in range(len(situation)):
+            if np.random.rand() > self.cfg.covering_wildcard_chance:
+                generalized.append(self.cfg.classifier_wildcard)
+            else:
+                generalized.append(situation[i])
+        cl = Classifier(cfg=self.cfg,
+                        condition=Condition(situation),
+                        action=action,
+                        time_stamp=time_stamp)
+        return cl
+
+    def _generate_covering_and_insert(self, situation, action, time_stamp):
+        cl = self.generate_covering_classifier(situation, action, time_stamp)
+        self.insert_in_population(cl)
+        self.delete_from_population()
+        return cl
+
+    # Roulette-Wheel Deletion
+    # TODO: use strategies
+    def delete_from_population(self):
+        if self.numerosity > self.cfg.max_population:
+            average_fitness = sum(cl.fitness for cl in self) / self.numerosity
+            deletion_votes = []
+            for cl in self:
+                deletion_votes.append(self._deletion_vote(cl, average_fitness))
+            selector = random.uniform(0, sum(deletion_votes))
+            self._remove_based_on_votes(deletion_votes, selector)
+
+    def _deletion_vote(self, cl, average_fitness):
+        vote = cl.action_set_size * cl.numerosity
+        if cl.experience > self.cfg.deletion_threshold and \
+                cl.fitness / cl.numerosity > \
+                self.cfg.delta * average_fitness:
+            vote *= average_fitness / (cl.get_fitness() / cl.numerosity)
+        return vote
+
+    # I toyed with numerosity -= 1
+    # and had better results with same solution as hosford42
+    # which is deleting the classifier
+    def _remove_based_on_votes(self, deletion_votes, selector):
+        for cl, vote in zip(self, deletion_votes):
+            selector -= vote
+            if selector <= 0:
+                assert cl in self
+                if cl.numerosity > 1:
+                    cl.numerosity -= 1
+                else:
+                    self.safe_remove(cl)
+                return None
+
+    def form_match_set(self, situation: Perception,  time_stamp):
+        matching_ls = [cl for cl in self if cl.does_match(situation)]
+        while len(matching_ls) < self.cfg.number_of_actions:
+            action = self._find_not_present_action(matching_ls)
+            cl = self._generate_covering_and_insert(situation, action, time_stamp)
```

Comment:
Podejrzany kod. Z jednej strony sygnatura (`form_match_set`) sugeruje że jakieś elementy zostaną wyfiltrowane. Z drugiej strony wewnętrzna lista jest modyfikowana w locie. Do przemyślenia - jak zwiększyć czytelność - bo to też może być miejsce gdzie ciężko znaleźć bład

Reply:
Zamieniłem nazwy `form_match_set` na `generate_match_set`, ale też zamieniłem `form_action_set` na `generate_action_set`.
Zależy mi by logika kodu, podział, i nazwy były jak najbliższe pseudokodu w An Algorithmic Description of XCS, więc po prostu zmieniam na nazwę tej metody by była taka sama jak w literaturze. Nie pamiętam dlaczego zdecydowałem się to nazwać inaczej, ale to był błąd z mojej strony.